common/src/main/java/net/caffeinemc/mods/lithium/common/block/redstone/RedstoneWirePowerCalculations.java

```diff
@@ -0,0 +1,222 @@
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.level.redstone.RedstoneWireEvaluator;
+
+/**
+ * Optimizing redstone dust is tricky, but even more so if you wish to preserve behavior
+ * perfectly. There are two big reasons redstone wire is laggy:
+ * <br>
+ * - It updates recursively. Each wire updates its power level in isolation, rather than
+ * in the context of the network it is a part of. This means each wire in a network
+ * could check and update its power level over half a dozen times. This also leads to
+ * way more shape and block updates than necessary.
+ * <br>
+ * - It emits copious amounts of duplicate and redundant shape and block updates. While
+ * the recursive updates are largely to blame, even a single state change leads to 18
+ * redundant block updates and up to 16 redundant shape updates.
+ * 
+ * <p>
+ * Unfortunately fixing either of these aspects can be detected in-game, even if it is
+ * through obscure mechanics. Removing redundant block updates can be detected with
+ * something as simple as a redstone wire on a trapdoor, while removing the recursive
+ * updates can be detected with locational setups that rely on a specific block update
+ * order.
+ * 
+ * <p>
+ * What we can optimize, however, are the power calculations. In vanilla, these are split
+ * into two parts:
+ * <br>
+ * - Power from non-wire components.
+ * <br>
+ * - Power from other redstone wires.
+ * 
+ * <p>
+ * For the default evaluator, we can combine the two to reduce calls to Level.getBlockState
+ * and BlockState.isRedstoneConductor as well as calls to BlockState.getSignal and
+ * BlockState.getDirectSignal. We can avoid calling those last two methods on redstone wires
+ * altogether, since we know they should return 0.
+ * <br>
+ * These changes can lead to a mspt reduction of up to 20% on top of Lithium's other
+ * performance improvements.
+ * 
+ * <p>
+ * The experimental evaluator actually uses the wire-power and non-wire-power separately,
+ * so we can't combine them. We can, however, optimize each individually, with the same
+ * principles.
+ * 
+ * @author Space Walker
+ */
+public class RedstoneWirePowerCalculations {
+
+    private static final int MIN = 0;            // smallest possible power value
+    private static final int MAX = 15;           // largest possible power value
+    private static final int MAX_WIRE = MAX - 1; // largest possible power a wire can receive from another wire
+
+    private static Block wireBlock;
+    private static RedstoneWireEvaluator evaluator;
+    private static boolean ignoreWires = false;
+    private static boolean ignoreNonWires = false;
+
+    public static int getNeighborBlockSignal(Block wireBlock, RedstoneWireEvaluator evaluator, Level level, BlockPos pos) {
+        ignoreWires = true;
```

Comment:
I am concerned about these static fields, mod compatibility with threading mods will definitely suffer

Reply:
I think converting them to a local variable should be possible if I am not misunderstanding the code

Reply:
yeah that'd work